MemoryPool: Allocate only once for preheating & impl. proper resetting.

two-horned opened the pull request:
Mempool calls the wrapped allocator N times when preheating it with N items.
I think it is more appropriate, if it only allocates one time. This should be faster and more importantly,
in case preheating fails (due to an allocation error), the state of the Mempool is easier to reason about.

two-horned committed:
Allocate only once for preheating.

More uniform comments.

Justus2308 commented:
I thought about this too in #23234, I'm not sure if there's any performance benefit to be had here though. The pool is already backed by an `ArenaAllocator` which batch allocates anyway so it probably won't make any difference in practice to ask for one large or for several small allocations. Asking for many smaller allocations might even allow the arena to batch allocate more optimally.

If you want, you can use this piece of code I stole from [the article about the original implementation of `MemoryPool`](https://zig.news/xq/cool-zig-patterns-gotta-alloc-fast-23h) to test if there's actually any performance benefit (you'd have to modify it slightly to use preheating):
<details>
<summary>code</summary>

```zig
pub fn main() !void {
    const gpa = std.heap.c_allocator;
    var rand: std.Random.DefaultPrng = .init(0);
    const r = rand.random();
    var open: std.ArrayList(*anyopaque) = try .initCapacity(gpa, 300);
    defer open.deinit(gpa);
    inline for (.{ [1]u8, [8 << 10]u8, [1 << 10 << 10]u8 }) |T| {
        var pool: std.heap.MemoryPoolExtra(T, .{}) = .init(gpa);
        defer pool.deinit();
        defer open.clearRetainingCapacity();
        for (0..2_500_000) |_| {
            const fill_rate = open.items.len / 256;
            const i = r.int(usize);
            if (i <= fill_rate) {
                pool.destroy(@ptrCast(@alignCast(open.swapRemove(i))));
            } else {
                const item = try pool.create();
                item[0] = 123;
                try open.append(gpa, @ptrCast(item));
            }
        }
    }
}

const std = @import("std");

```

</details>

Just a warning, running this benchmark *can* trigger the OOM killer (at least on my machine), if it does just adjust the allocation sizes until it doesn't.

Maybe a good solution would be to make `preheat` return a `usize` count of how many nodes were actually allocated, that way you're still able to preheat optimistically until you hit OOM *and* it's easier to reason about how many nodes are currently available.

two-horned committed:
Implement TODO: Resets for mempools.

Respect arena reset result.

two-horned commented:
I have just implemented the TODO. Resetting a Memory pool is now possible.

The modes `free_all` and `retain_capacity` work similar to the arena allocator versions, however `retain_with_limit` takes a limit of n `Item` instances (in contrast to n bytes).

Why don't we need to save a list of nodes? Because building nodes is actually pretty cheap with the new `preheat` solution: only one allocation call and concatenating the nodes. The allocation is also guaranteed to succeed (if the arena has succeeded in resetting).

Added basic test to showcase the capabilities.

two-horned committed:
Fix typos and naming.

two-horned committed:
Fix typos, remove unused function

two-horned committed:
Fix reset statements, needs allocator

two-horned committed:
Fix reset.

two-horned committed:
Another fix. Writing 'allocator' everywhere is hard...

two-horned committed:
Release type dependency for MemoryPool. Fix tests.

two-horned committed:
Revert importing "std" instead "../std.zig".

alexrp commented:
Closing because we have no CI on GitHub anymore and this has no CI results.

Please reopen this on Codeberg: https://codeberg.org/ziglang/zig/pulls
